Content.Server/Power/Pow3r/BatteryRampPegSolver.cs

```diff
@@ -115,11 +115,10 @@ private void UpdateNetwork(Network network, PowerState state, float frameTime)
 
                 var batterySpace = (battery.Capacity - battery.CurrentStorage) * (1 / battery.Efficiency);
                 batterySpace = Math.Max(0, batterySpace);
-                var scaledSpace = batterySpace / frameTime;
 
-                var chargeRate = battery.MaxChargeRate + battery.LoadingNetworkDemand / battery.Efficiency;
+                var chargeRate = Math.Min(battery.MaxChargeRate, batterySpace) + battery.LoadingNetworkDemand / battery.Efficiency;
 
-                battery.DesiredPower = Math.Min(chargeRate, scaledSpace);
+                battery.DesiredPower = chargeRate;
```

Comment:
I suspect this line is doing all of the heavy lifting (fixing the power flickering).

In the old code, the pass through load `battery.LoadingNetworkDemand / battery.Efficiency` is limited by the battery charge rate `scaledSpace`, whereas in the new code it is not.

Reply:
I appreciate the clarification in the previous comments and my changes were somewhat uninformed since this is has been my first foray into the code base. Perhaps I spent too long trying to figure out how it worked and didn't pay enough attention to what the variable names were, haha. In the middle of typing this response after your clarifications I realized why some things were the way they were. I'll look into revising the code so it makes more sense and try to clear up the unit mismatches.

Reply:
Since you mention that the charge rate limitation, _should_ the pass through load be limited by the charge rate? If I recall correctly the comments in the code mention potentially changing that but the way I did it is definitely a bit "hacky."

Reply:
I'm currently putting together a local test setup to see if this line is actually what is affecting this. What would actually help me most here is a saved test map, e.g. one that I could load in with a minimum amount of setup to see that the power is blinking before the fix and see that power isn't blinking after the fix.

I've set up a local Bagel with a singularity but haven't gotten any power blinks yet.
(though I distinctly remember this happening quite often in actual games)

Reply:
Maybe it's my memory failing but I can't remember if the flickers are station-wide or only certain APCs, but if it's the latter it could be only certain APCs that have a higher than average draw than others. The medbay on Box for example has a relatively large draw so when the supply drops suddenly, say inbetween tesla coil shocks, maybe the battery just can't ramp up fast enough?

Reply:
> I've set up a local Bagel with a singularity but haven't gotten any power blinks yet.
(though I distinctly remember this happening quite often in actual games)

you fixed low-fuel rad collectors flickering power already, the issue happens even with just ame -> smeses -> substations

Reply:
I've confirmed through more testing that most of the changes involving my unit mismatches weren't changing much, and have since been reverted, but there were 2 lines that prevented the flickering and it is basically what you said @Partmedia.

var chargeRate = Math.Min(battery.MaxChargeRate, batterySpace);
battery.DesiredPower = chargeRate + battery.LoadingNetworkDemand / battery.Efficiency;

While batterySpace is energy and MaxChargeRate is in power, the limit is to make sure we never pull more than the battery has capacity for in a single tick rather than over one second. I tried making it scale to be the same units but then there is a massive overdraw of power over the network. Most of the time batteries are only missing a few joules of power but are requesting a draw in watts so they end up always requesting the Max Charge because what they are missing is usually 273-ish joules but in watts it is well over 10000 joules.

The DesiredPower is now also guaranteed to request the network load and add what is needed for the battery on top of that. Through my testing I did observe that power flickers as before while generators are ramping up/down and the SMES/Substations ramp to compensate, but this seems like intended design. The consistent flickering due to fluctuations in the generator supply after equalization on the grid seem to be absent.

Reply:
If it's okay with you, let's go ahead and merge this PR with changes to just these two lines. If you have trouble cleaning it up in the repo side on your end, I can help you with that.

Your unit fixing changes below are appreciated, but I just want to make sure those go into a separate PR so that they don't affect the meat of the changes here.

Reply:
Just to be clear, are you asking me to open a new PR that only has changes to those two lines? If so I can do that pretty quick, after I test it again of course.

Reply:
I was asking for a new PR with the other changes, so that this PR only contains the two line change. But the opposite is fine, too. I just want these two lines in a separate PR so that it is clear what is actually doing the power blink fixing.

Reply:
Ah gotcha, I have a backup of the initial PR saved so I'll just make a new local branch with that one and create a PR for it.

Reply:
I decided to upload the revised as a new PR instead of the original so this can remain as is since it has the original changes anyway and ran into a snag since I started from the master file and just change the two lines, but it was just a Debug.Assert checking to make sure the power being added to the battery was within a small margin of the max capacity right before being clamped anyway. For now I commented that out with an explanation, but I've also observed that due to my changes the AME continues to ramp up until it reaches the targeted power supply instead of stopping around the load for some reason. The TEG does this as well. As such it, along with the other engines, being passed through the batteries means that touching high voltage cables is *incredibly* dangerous when supplied by generators. This was not intended and I'm looking into finding a better solution. Maybe it's best to close this PR for now?